app/cloud.go

```diff
@@ -18,6 +21,59 @@ func (a *App) getSysAdminsEmailRecipients() ([]*model.User, *model.AppError) {
 	return a.GetUsers(userOptions)
 }
 
+// SendAdminUpgradeRequestEmail takes the username of user trying to alert admins and then applies rate limit of n (number of admins) emails per user per day
+// before sending the emails.
+func (a *App) SendAdminUpgradeRequestEmail(username string, subscription *model.Subscription) *model.AppError {
+	if a.Srv().License() == nil || (a.Srv().License() != nil && !*a.Srv().License().Features.Cloud) {
+		return nil
+	}
+
+	if subscription != nil && subscription.IsPaidTier == "true" {
+		return nil
+	}
+
+	if a.Srv().EmailService.PerDayEmailRateLimiter == nil {
+		return model.NewAppError("app.SendAdminUpgradeRequestEmail", "app.email.no_rate_limiter.app_error", nil, fmt.Sprintf("username=%s", username), http.StatusInternalServerError)
```

Comment:
Why don't use the params parameter of NewAppError to include the username?

Reply:
Our translation id doesn't have any params included
```json
  {
    "id": "app.email.no_rate_limiter.app_error",
    "translation": "Rate limiter is not set up."
  }

```
In this case the username is just used by AppError's ```Error``` method

Reply:
Yeah I mean, Why don't use `map[string]interface{}{"username": username}` in the parameters param and `err.Error()` in the error part of the `AppError` constructor?

Reply:
Something like this?

```go
	if a.Srv().EmailService.PerDayEmailRateLimiter == nil {
		return model.NewAppError("app.SendAdminUpgradeRequestEmail", "app.email.no_rate_limiter.app_error", map[string]interface{}{"username": username}, fmt.Sprintf("for username=%s", username), http.StatusInternalServerError)
	}
```

Reply:
Oh, didn't see that this comes from a nil comparison. I'll use something like:

```
if a.Srv().EmailService.PerDayEmailRateLimiter == nil {
		return model.NewAppError("app.SendAdminUpgradeRequestEmail", "app.email.no_rate_limiter.app_error", map[string]interface{}{"username": username}, "", http.StatusInternalServerError)
	}
```

because it doesn't come from an error

Reply:
Doing it like that would give us an error like so ```"detailed_error": "app.SendAdminUpgradeRequestEmail: Rate limiter is not set up., ",``` that's missing the username instead of the original that would be ```"detailed_error": "app.SendAdminUpgradeRequestEmail: Rate limiter is not set up., for username=test2",```

4/5 we leave it as it is

Reply:
No hard thoughts on that. I think that if the constructor has the params it should appear in the logs, probably something is not in place but not important though. Resolving this one